feat: support env var in varnish config for servers key

tft7000 opened the pull request:
I tried to solve #521 with this pull request:

in addition to the regular case (array of servers) the configuration for `varnish.servers` can now be a string holding an environment variable. The environment variable is supposed to be a string that contains a json array of servers.

This solves the problem, that the same app under different environments might have different server configurations. Eg some test setup might have one varnish and a production setup might have three varnish.

To solve the issue, I had to loosen the checks in `Configuration.php` a bit.

dbu commented on src/DependencyInjection/Configuration.php:
```diff
@@ -531,12 +531,9 @@ private function getHttpDispatcherNode()
-                ->arrayNode('servers')
+                ->variableNode('servers')
                     ->info('Addresses of the hosts the caching proxy is running on. May be hostname or ip, and with :port if not the default port 80.')
-                    ->useAttributeAsKey('name')
```
afaik this is breaks xml configuration.

i wonder if it would be best to make this node not required and add a separate node for the env variable and have the Configuration class check that exactly one of the two nodes is set. and in the extension pick the one we have.

that would not mess with xml or the array validation.

dbu commented on src/DependencyInjection/FOSHttpCacheExtension.php:
```diff
@@ -365,12 +365,22 @@ private function loadProxyClient(ContainerBuilder $container, XmlFileLoader $loa
+        if (is_string($config['servers'])) {
             $usedEnvs = [];
-            $container->resolveEnvPlaceholders($url, null, $usedEnvs);
+            $container->resolveEnvPlaceholders($config['servers'], null, $usedEnvs);
```
does this allow to specify a list of urls in a single env variable? or should we split on `,`? (though `,` is technically a legal part of a url, in practice this is a list of ips or host names, so it would be strange to have paths with `,` in it)

dbu reviewed (changes requested):
thanks for picking this topic up again.

what do you think of my suggestion to split this into 2 separate configuration fields to keep validation doable?

shame that the build seems to have deteriorated, does not seem related to your changes. we should fix that build issue separately, but i am very busy atm... worst case i think we can merge when some builds are green to not block this unnecessarily.

dbu commented on tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php:
```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
+
+        $config = $this->getBaseConfig();
+        $config['proxy_client']['varnish']['http']['servers'] = $container->getParameter('triggerServersValue');
+        $config['proxy_client']['varnish']['http']['servers_from_jsonenv'] = $container->getParameter('triggerServersFromJsonEnvValue');
```
```suggestion
        if (null !== $serversFromJsonEnvValue) {
            $config['proxy_client']['varnish']['http']['servers_from_jsonenv'] = $container->getParameter('triggerServersFromJsonEnvValue');
        }
```

dbu commented on tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php:
```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
+        (new ResolveParameterPlaceHoldersPass())->process($container);
+
+        $config = $this->getBaseConfig();
+        $config['proxy_client']['varnish']['http']['servers'] = $container->getParameter('triggerServersValue');
```
we should only configure the option if we are supposed to configure it. then the `requiresAtLeastOneElement` works
```suggestion
        if (null !== $serversValue) {
            $config['proxy_client']['varnish']['http']['servers'] = $container->getParameter('triggerServersValue');
        }
```

tft7000 replied:
thanks - in addition, I need to unset the servers in the else case, as it is defined in `getBaseConfig`

dbu commented on tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php:
```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
+        $_ENV['VARNISH_SERVERS'] = $envValue;
+        $container = $this->createContainer();
+
+        // workaround to get the possible env string into the EnvPlaceholderParamterBag
```
```suggestion
        // workaround to get the possible env string into the EnvPlaceholderParameterBag
```

dbu commented on tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php:
```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
+     */
+    public function testVarnishServersConfig($serversValue, $serversFromJsonEnvValue, $envValue, $expectedServersValue, $expectExceptionClass, $expectExceptionMessage)
+    {
+        if ($expectExceptionClass) {
```
the exceptions should be thrown on the line with `$this->extension->load([$config], $container);`, right? lets move it to right before that line to not get confusing reports if an exception is thrown before then.

tft7000 replied:
you mean, in case the test preparation is throwing an exception? ok.

dbu commented on tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php:
```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
+        $this->extension->load([$config], $container);
+        (new ResolveEnvPlaceholdersPass())->process($container);
+
+        $definition = $container->getDefinition('fos_http_cache.proxy_client.varnish.http_dispatcher');
```
if we expect an exception, we should not get here. lets check for $expectExceptionClass before this again and return if we expected an exception. while trying stuff, i made the exception not be thrown and then got a confusing test failure on line 700.

dbu commented:
i fixed the build in the master branch. if you can rebase your branch on upstream master, the CI should run through again.

tft7000 commented on src/DependencyInjection/FOSHttpCacheExtension.php:
```diff
@@ -372,6 +372,15 @@ private function createHttpDispatcherDefinition(ContainerBuilder $container, arr
                 $this->validateUrl($url, 'Not a valid Varnish server address: "%s"');
             }
         }
```
It seems that an ArrayNode will always lead to an array, so I need to manually fix that in the validation method (similar to the validation for `default_ban_headers`).

dbu replied:
thanks

tft7000 commented:
hmm, sorry, I messed up with the merge...

tft7000 commented:
@dbu thank you very much for your detailed feedbacks! I changed the code accordingly and updated the docu and changelog.

dbu commented on src/DependencyInjection/FOSHttpCacheExtension.php:
```diff
@@ -365,12 +365,23 @@ private function loadProxyClient(ContainerBuilder $container, XmlFileLoader $loa
     private function createHttpDispatcherDefinition(ContainerBuilder $container, array $config, $serviceName)
     {
-        foreach ($config['servers'] as $url) {
+        if (isset($config['servers'])) {
```
```suggestion
        if (array_key_exists('servers', $config)) {
```

dbu commented on src/DependencyInjection/FOSHttpCacheExtension.php:
```diff
@@ -365,12 +365,23 @@ private function loadProxyClient(ContainerBuilder $container, XmlFileLoader $loa
+                }
+            }
+        }
+        if (isset($config['servers_from_jsonenv']) && is_string($config['servers_from_jsonenv'])) {
```
```suggestion
        if (array_key_exists('servers_from_jsonenv', $config) && is_string($config['servers_from_jsonenv'])) {
```

dbu commented:
thanks a lot! i added some nitpicks (i prefer array_key_exists over isset, because isset does not error if the variable name of the array does not exist due to code changes or typos)

can you try to squash your commits and rebase on master again? with `git rebase -i upstream/master` (or whatever this repo is called in your git) you should be able to select which commits to keep and which to squash. the ones from me should not show up in the MR, or they would end up being duplicated ;-)

dbu commented:
thanks, the change looks good now!

do you need help with the git cleanup? i can also do that and merge on the cli if you prefer.

tft7000 commented:
thanks - actually I would really like to know how to fix my merge mess. so if you can give me a pointer that would be great ;-) However, if you are short on time, i very much understand that.

I got a remote-tracking branch merge when I pushed todays fixes. somehow with rebasing and sqashing I get not over that - I just create new remote tracking branches..

dbu commented:
if i would want to solve the current state, i would do

    git rebase -i origin/master

and then change `pick` to `s` for all your commits (except the first one) and delete the lines with the changes that come from master, so that those changes get dropped. want to try?

tft7000 commented:
I did try that (several times): I do get (almost) only my commits:
```
pick 2c1b76e feat: support env var in varnish config for servers key
pick 0c6b762 feat: added servers_from_jsonenv config node that overwrites servers config
pick 44cd412 fix: fixes based on review, added docu
pick 2830820 feat: new config options servers_from_jsonenv to support a variable amount of proxy servers defined via an env var
pick b0e70f4 feat: support env var in varnish config for servers key
pick 0aea4a2 feat: added servers_from_jsonenv config node that overwrites servers config
pick 57d4aa7 fix: fixed style
pick f8afc97 fix: fixed messed up merge
pick bff1363 Update src/DependencyInjection/Configuration.php
pick eedb8ff refactor: use array_key_exists instead of isset
```
when I am thru with all the conflicts I try to squash the commits - however I never seem to get past a 'merge remote-tracking branch' - I cannot squash my commits from last week. I don't know.. I seem to miss some basics...

![image](https://user-images.githubusercontent.com/3369340/131354434-86cb2daa-45a0-4fee-824f-e20930aed02b.png)

tft7000 commented:
@dbu sorry for the delay.. i hope the PR is now ok.

dbu commented on Resources/doc/reference/configuration/proxy-client.rst:
```diff
@@ -38,6 +38,7 @@ varnish
                         - 123.123.123.1:6060
                         - 123.123.123.2
                     base_url: yourwebsite.com
+                    servers_from_jsonenv: '%env(json:VARNISH_SERVERS)%'
```
```suggestion
                    # alternatively, if you configure the varnish servers in an environment variable:
                    # servers_from_jsonenv: '%env(json:VARNISH_SERVERS)%'
```

dbu commented on Resources/doc/reference/configuration/proxy-client.rst:
```diff
@@ -69,6 +70,24 @@ When using a multi-server setup, make sure to include **all** proxy servers in
 this list. Invalidation must happen on all systems or you will end up with
 inconsistent caches.
 
+Note: when using a variable amount of proxy servers that are defined via environment
```
```suggestion
.. note::

     When using a variable amount of proxy servers that are defined via environment
```

dbu commented on Resources/doc/reference/configuration/proxy-client.rst:
```diff
@@ -38,6 +38,7 @@ varnish
                         - 123.123.123.1:6060
                         - 123.123.123.2
                     base_url: yourwebsite.com
```
```suggestion
```

dbu commented on Resources/doc/reference/configuration/proxy-client.rst:
```diff
@@ -69,6 +70,24 @@ When using a multi-server setup, make sure to include **all** proxy servers in
 inconsistent caches.
 
+Note: when using a variable amount of proxy servers that are defined via environment
+variable, use the ``http.servers_from_jsonenv`` option below.
```
```suggestion
    variable, use the ``http.servers_from_jsonenv`` option below.
```

dbu commented:
thanks a lot!

released as https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/releases/tag/2.11.0
